feat(txpool): add price bump for `is_underpriced` transaction check

tcoratger opened the pull request:
In go-ethereum implementation, the verification for transaction underpricing includes a price bump following the formula: `gas * (100 + priceBump) / 100`. 

In the default configuration, price bump is 10 %: https://github.com/ethereum/go-ethereum/blob/73697529994e14996b7740730481e926d5ec3e40/core/txpool/txpool.go#L188.

As a result, in this PR, a 10% price bump constant is proposed and implemented inside the `is_underpriced` transaction check.

Should solve #2958.

mattsse commented on crates/transaction-pool/src/validate.rs:
```diff
@@ -369,7 +369,8 @@ impl<T: PoolTransaction> ValidPoolTransaction<T> {
     /// Returns true if this transaction is underpriced compared to the other.
     pub(crate) fn is_underpriced(&self, other: &Self) -> bool {
-        self.transaction.effective_gas_price() <= other.transaction.effective_gas_price()
+        self.transaction.effective_gas_price() * (100 + PRICE_BUMP) / 100 <=
+            other.transaction.effective_gas_price()
```
I believe we need to change this a bit so that this matches:

https://github.com/ethereum/go-ethereum/blob/d40a255e973775575d8d16456252f93ac75c09f0/core/txpool/legacypool/list.go#L288-L306

which looks a bit more complex than the existing check

tcoratger replied:
@mattsse Done I've mimic their test in order to check if:
- Transaction A max fee per gas is less that Transaction B max fee per gas + a price bump.
- Or Transaction A max priority fee per gas is less that Transaction B max priority fee per gas + a price bump (in this situation, the two max priority fee per gas should not be zero indicating a non-EIP1559 transaction and invalidating the test).
- Or Transaction A effective gas price is less than Transaction B effective gas price.

mattsse reviewed (changes requested):
ty for taking this on.

I'd also appreciate a simple test, similar to https://github.com/paradigmxyz/reth/blob/main/crates/transaction-pool/src/pool/txpool.rs#L1390-L1390

tcoratger commented on crates/transaction-pool/src/validate.rs:
```diff
@@ -369,7 +369,8 @@ impl<T: PoolTransaction> ValidPoolTransaction<T> {
-        self.transaction.effective_gas_price() <= other.transaction.effective_gas_price()
+        self.transaction.effective_gas_price() * (100 + PRICE_BUMP) / 100 <=
+            other.transaction.effective_gas_price()
     }
```
@mattsse This is done.

tcoratger commented:
@mattsse For your information:
- I've added a simple test for replace transaction with underpricing.
- I've removed the `self.transaction.effective_gas_price() <= other.transaction.effective_gas_price()` condition which is replaced in my opinion by the conditions on `max_fee_per_gas` and `max_priority_fee_per_gas` (otherwise it's redundant because `effective_gas_price` contains itself the `max_fee_per_gas` and `max_priority_fee_per_gas` items). In the Go implementation there is no condition on `effective_gas_price`.
- For non-EIP1559 transactions: `max_priority_fee_per_gas` is zero and then only condition on `max_fee_per_gas` is taken into account.

onbjerg reviewed (approved):
looks good to me, pending @mattsse

mattsse reviewed (approved):
awesome tysm,

love the test
